### PR TITLE
1-clean up retrofit dependency

### PR DESCRIPTION
**Pull request by lnt-nishantdande**
@nets-dev As per review comments, I have removed retrofit dependency from the API.kt file. Added new NetworkModule.kt file to handle all retrofit configuration and services. Update NetworkModule class function calls in RetrofitAPI.kt file.

### CODE REVIEW

**Comment by lnt-nishantdande** [comment]
++ @nets-dev

File: app/src/main/java/com/example/practice/MainActivityViewModel.kt
Comment (nets-dev) [inline app/src/main/java/com/example/practice/MainActivityViewModel.kt:38]: The initialisation `API.create()` is executed each time this function is invoked. Meaning each network fetch is handled by a different instance of `API` implementation. 

1. What is the advantage and disadvantage of this approach?
2. What happens in case of multiple simultaneous calls (e.g. if we call this function from a *Refresh* button action)?
3. What makes each instance stay in memory after the function scope has exited? 
4. Is it possible the application can struggle to garbage collect these instances? If yes, in what scenario? If no, how can we be sure that will not happen? (Give straightforward and concrete scenarios/reasons)

(Note that the answers, here and all other questions, should target the question directly as the idea is to discuss programming fundamentals. Therefore, feel free to ask if anything is unclear before answering the questions)

- [ ] Are all these questions clear? If not, comment your question below

File: app/src/main/java/api/NetworkModule.kt
Comment (nets-dev) [inline app/src/main/java/api/NetworkModule.kt:28]: Singletons are generally discouraged in programming. What makes the use of a singleton object here the right choice? 
Also discuss the difference between `object SomeObject {}` and `class SomeClass {}` and where one should be used over the other below.

File: app/src/main/java/api/NetworkModule.kt
Comment (nets-dev) [inline app/src/main/java/api/NetworkModule.kt:22]: What is the difference between `Deferred` for this async request in comparison to `suspend` coroutine approach?

File: app/src/main/java/api/RetrofitAPI.kt
Comment (nets-dev) [inline app/src/main/java/api/RetrofitAPI.kt:19]: - This function executes network fetch inside `runBlocking`. What does run-blocking mean and why is it the right choice here? 
- Upon receiving response from server, `failure` with `IllegalArgumentException` is invoked when no data is received. Is this the right exception to throw? (Refer to IllegalArgumentException documentation)

File: app/src/main/java/com/example/practice/MainActivityViewModel.kt
Comment (nets-dev) [inline app/src/main/java/com/example/practice/MainActivityViewModel.kt:41]: 1. The success callback if `fetchUsers` is first handled inside `viewModelScope` using IO dispatch. Then execution is moved to `Main` dispatch to update `_userDataState`. What is the reason for each choice of dispatch? 
i.e. What is achieved by dispatching to IO, and what is achieved by dispatching from IO to Main afterwards?
2. In contrary, failure callback is not dispatched. Why is that ok? And what thread will failure be executing on considering its not explicitly dispatched to any thread?

Comment (nets-dev) [inline app/src/main/java/com/example/practice/MainActivityViewModel.kt:41]: Line 78 contains the following statement: `it?.id?.equals(excludingUserId)!!`
What scenario could cause application crash at this line? How would you avoid it?

**Review by nets-dev — CHANGES_REQUESTED** [review]
Please see each review comment and **reply to each** question. Feel free to ask if anything is unclear in the questions.

File: app/src/main/java/com/example/practice/MainActivityViewModel.kt
Comment (lnt-nishantdande) [reply app/src/main/java/com/example/practice/MainActivityViewModel.kt:41]: @nets-dev code will be expected to crash when 'id' will set to 'null'.
Fix : 
Check whether 'id' is null and if 'id' is null return 'false' or return 'id.equals(excludingUserId)'

Comment (lnt-nishantdande) [reply app/src/main/java/com/example/practice/MainActivityViewModel.kt:41]: @nets-dev
Q1] The success callback if fetchUsers is first handled inside viewModelScope using IO dispatch. Then execution is moved to Main dispatch to update _userDataState. What is the reason for each choice of dispatch?
    i.e. What is achieved by dispatching to IO, and what is achieved by dispatching from IO to Main afterwards?
    **Ans** :- 1. The purpose of using Dispacter.IO is that we are performing excluding operations that are time-consuming. So if I did not use Dispacter.IO then we might face an ANR issue due to blocking UI thread.
                2. By using Dispacther.IO we perform a long time taking operations smoothly on a different thread and once the operation gets complete dispatch it on the Main thread using Dispatcher.Main.

Q2] In contrary, failure callback is not dispatched. Why is that ok? And what thread will failure be executing on considering its not explicitly dispatched to any thread?
    **Ans** :- fallback callback is not dispatched because it is handled in the main thread so no need to dispatch separately on the main thread.

File: app/src/main/java/api/RetrofitAPI.kt
Comment (lnt-nishantdande) [reply app/src/main/java/api/RetrofitAPI.kt:19]: @nets-dev

Q1] This function executes network fetch inside runBlocking. What does run-blocking mean and why is it the right choice here?
**Ans** :-  As the network call is async, hence we have to use await() to get the API result. As await() is a suspended function, it has to handle inside coroutine i.e run-blocking/launch. run-blocking is a coroutine builder which waits for completion. 
    
Q2] Upon receiving response from server, failure with IllegalArgumentException is invoked when no data is received. Is this the right exception to throw? (Refer to IllegalArgumentException documentation).
**Ans** :- IllegalArgumentException is used because whenever we request a server with an illegal parameter we are getting a bad request or null error. Here in this case if I am getting null or bad requests so it should specifically mention error due to Illegal arguments.

File: app/src/main/java/api/NetworkModule.kt
Comment (lnt-nishantdande) [reply app/src/main/java/api/NetworkModule.kt:22]: @nets-dev
Q1] What is the difference between Deferred for this async request in comparison to suspend coroutine approach?
**Ans** :- In Deferred, as it is the async request we have to wait for the result using await(). But suspend approach make this thing super easy. I should use suspend instead of Deferred. Thank you for pointing this approach to me.

File: app/src/main/java/api/NetworkModule.kt
Comment (lnt-nishantdande) [reply app/src/main/java/api/NetworkModule.kt:28]: @nets-dev
Q1] Singletons are generally discouraged in programming. What makes the use of a singleton object here the right choice?
**Ans** :- As user service API is a frequently used service, hence making a singleton object is the best choice to implement this scenario. If I make use of 'class' then for each and every request I have to create an instance and access user service API. To avoid memory overhead I have to make NetworkModule a singleton object.
Q2] Also discuss the difference between object SomeObject {} and class SomeClass {} and where one should be used over the other below.
**Ans** :- object - The 'object' is used when we have to implement a singleton. We get only a single reference throughout the app and can access member variable and functions with the same reference.
                 class - we can create multiple instances of the class. We need to create an instance to access member variables and functions.

File: app/src/main/java/com/example/practice/MainActivityViewModel.kt
Comment (lnt-nishantdande) [reply app/src/main/java/com/example/practice/MainActivityViewModel.kt:38]: @nets-dev

Q1] What is the advantage and disadvantage of this approach?
**Ans** :- Advantage - 1. It hides the business logic of the component.
        			      2. It is loosely coupled with components. So that we can easily replace our business logic with new approach.
			              3. Better to maintain code structure.
          Disadvantage -   1. We have to take care that API.create() should return the same instance for each network call otherwise garbage memory increases.

Q2] What happens in case of multiple simultaneous calls (e.g. if we call this function from a Refresh button action)?
**Ans** :- RetrofitAPI instance will be get created multiple times and garbage collection will increase.

Q3] What makes each instance stay in memory after the function scope has exited?
**Ans** : - Yes. Instances stay in memory. It is because, in API.create() function we are returning a new instance of RetrofitAPI class every time.

Q4] Is it possible the application can struggle to garbage collect these instances? If yes, in what scenario? If no, how can we be sure that will not happen? (Give straightforward and concrete scenarios/reasons)
**Ans** :- Yes, it is possible applications can struggle to garbage collect these instances. Whenever we make a new api request, a new RetrofitAPI instance is created which leads to an increase in garbage memory. To avoid it we can return the single instance of RetrofitAPI from API.create() function.

Comment (nets-dev) [reply app/src/main/java/com/example/practice/MainActivityViewModel.kt:38]: @lnt-nishantdande Thank you for the answers! I think there is a misunderstanding on question 3 so I'll explain further. Local variables are freed when a function scope exits, e.g. 
```
fun printHelloWorld() {
    val helloWorld: String = "Hello World!"
    print(helloWorld)
}
```
The local property `helloWorld` is freed at the end of the function scope, in other words, after the closing bracket - **}**. That means, the local property obtained from `API.create()` should normally be freed as soon as `fetchUserList` function reaches end of function scope at line 62. 
So the question is, what makes this instance stay in memory way beyond the function scope and execute a callback upon networking response. 
Is the question now clear?

File: app/src/main/java/com/example/practice/MainActivityViewModel.kt
Comment (nets-dev) [reply app/src/main/java/com/example/practice/MainActivityViewModel.kt:41]: @lnt-nishantdande About the second question, what guarantees that failure callback will be invoked on main thread?

File: app/src/main/java/api/NetworkModule.kt
Comment (nets-dev) [reply app/src/main/java/api/NetworkModule.kt:22]: You're welcome! There are a couple of comments on your answers below which you can take a look and reply. 
In the meantime, we'd like to setup a meeting with you tomorrow. Are you available 11 - 12:00 Finnish time? 
Please let us know below and we'll send you a meeting link. (**Note that the time is Finnish time**)

Meeting topics will be:

- Discussion on the project and comments in this review process 
- Brief overview of what our projects are and what sort of skills we find important 
- Discussion on what resources you've used while working on this project and any documentation, blog or website you found helpful 

How does that sound? @lnt-nishantdande

**Review by nets-dev — CHANGES_REQUESTED** [review]

**Comment by nets-dev** [comment]
There are a couple of comments on your answers below which you can take a look and reply.
In the meantime, we'd like to setup a meeting with you tomorrow. Are you available 11 - 12:00 Finnish time?
Please let us know below and we'll send you a meeting link. (Note that the time is Finnish time)

Meeting topics will be:

Discussion on the project and comments in this review process
Brief overview of what our projects are and what sort of skills we find important
Discussion on what resources you've used while working on this project and any documentation, blog or website you found helpful
How does that sound? @lnt-nishantdande

File: app/src/main/java/api/NetworkModule.kt
Comment (lnt-nishantdande) [reply app/src/main/java/api/NetworkModule.kt:22]: @nets-dev Yes, I will be available between 11 -12:00 Finnish time i.e 1:30 PM - 2:30 PM IST.

Comment (nets-dev) [reply app/src/main/java/api/NetworkModule.kt:22]: @lnt-nishantdande Great, here is the teams link for the meeting:
[Teams meeting 11 - 12:00](https://teams.microsoft.com/l/meetup-join/19%3ameeting_YmM1ZmRkY2UtOWZiNS00MjI0LWE2YzEtYjk4MGNmZmVhMDAw%40thread.v2/0?context=%7b%22Tid%22%3a%2279dc228f-c8f2-4016-8bf0-b990b6c72e98%22%2c%22Oid%22%3a%2213eab4a7-c748-4cb4-809e-27977c6812e2%22%7d)

File: app/src/main/java/com/example/practice/MainActivityViewModel.kt
Comment (lnt-nishantdande) [reply app/src/main/java/com/example/practice/MainActivityViewModel.kt:41]: @nets-dev For question 2, as api call is made from run-blocking without using any dispatcher/CoroutineScope. So run-blocking will run async API requests and wait for a response. Once the result/error is received, pass it to the success/failure function which is in the main thread. Hence it is guaranteed that a failure callback will be invoked on the main thread.

File: app/src/main/java/com/example/practice/MainActivityViewModel.kt
Comment (lnt-nishantdande) [reply app/src/main/java/com/example/practice/MainActivityViewModel.kt:38]: @nets-dev Thank you for explaining me the Q3 in detail.
According to me, as we are calling create function statically means create function is placed in companion object { } and hence instance stay in memory till application lifecycle.

**Review by nets-dev — APPROVED** [review]
Thank you for the nice session! I will create the second issue like we discussed to:

- Learn Jetpack Compose list updating mechanism and implement it 
- Improve some aspects of the project from our discussions (as much as you can remember)

As always, feel free to reach out if you have any question!

**Comment by nets-dev** [comment]
I have approved this PR but I do not have write access to merge it so feel free to do that yourself @lnt-nishantdande

**Comment by lnt-nishantdande** [comment]
@nets-dev I enjoyed the discussion with you. I will improve the codebase as per our discussion and update you on the same.
Lastly, I will merge this branch.

**Comment by lnt-nishantdande** [comment]
> Thank you for the nice session! I will create the second issue like we discussed to:
> 
>     * Learn Jetpack Compose list updating mechanism and implement it
> 
>     * Improve some aspects of the project from our discussions (as much as you can remember)
> 
> 
> As always, feel free to reach out if you have any question!

@nets-dev 
As per our discussion I have updated the codebase. Below are the points which I have covered and modified the code.

1. Replace async/await network call with suspend.
2. No need to use 'IllegalArgumentException’ instead use appropriate exception as per server documentation. As in our case api documentation is not available I have used ‘Exception’ for now.
3. Removed multiple api calls from NetworkModule class and made a simple api to get user api service.
4. For the list updating mechanism, we have used 'observeAsState' to maintain the state of the UI. There is no need to add an extra mechanism to remember the list state. We have the requirement to re-fetch data every time when we select the row. So data already gets an update in LiveData which we observe it as a state. 
    Also, If we have to maintain data without performing the re-fetch operation, we can use ‘remember’ with a mutable state list and update remember state list item accordingly.

I have created a new branch feature/improvement, please do a review and provide a comment on the same. 
Codebase Link : -
https://github.com/lnt-nishantdande/Practice-Android/tree/feature/improvement

**Pull request merged**
